### PR TITLE
feat: Add socketTimeout to HTTP client config for closing stale connections

### PR DESCRIPTION
**Pull request by sichanyoo**
## Issue \#

https://github.com/awslabs/aws-sdk-swift/issues/1349

Companion PR: https://github.com/awslabs/aws-sdk-swift/pull/1413

## Description of changes

- Adds a new config `socketTimeout`, which gets used to construct `HTTPMonitoringOptions` for `CRTClientEngine`. Sets the maximum time to wait between data packets, which is used to close stale connections.
- Fixes `URLSessionHTTPClient` to use `socketTimeout` for its `timeoutIntervalForRequest` config instead of `connectTimeout`. Refer to this [official doc form Apple](https://developer.apple.com/documentation/foundation/nsurlsessionconfiguration/1408259-timeoutintervalforrequest), which confirms that `timeoutIntervalForRequest` maps to the new `socketTimeout` rather than `connectTimeout`, which is a connection timeout config.
- Adds missing `continuation.resume()` that makes the application hang when socket timeout error is thrown.
- Misc. comment change for clarification on what `connectTimeout` actually is for. It's not a socket timeout for limiting stale connections. It's used to limit time it takes to obtain a live connection to the server.

## Prerequisite
- Blocked on the aws-crt-swift PR: https://github.com/awslabs/aws-crt-swift/pull/242, which makes initializer for `HTTPMonitoringOptions` public for usage in smithy-swift.
- Blocked on bumping aws-crt-swift version to the new release that contains changes from above PR.

## Testing
Tested against localhost that mocks DynamoDB & does not return a response. HTTP client connection now closes and recovers from no server response as it should, instead of hanging for eternity.

## Scope

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Sources/ClientRuntime/Networking/Http/CRT/CRTClientEngine.swift [hunk 1]
```diff
@@ -338,6 +346,7 @@ public class CRTClientEngine: HTTPClient {
                 response.statusCode = makeStatusCode(statusCode)
             case .failure(let error):
                 self.logger.error("Response encountered an error: \(error)")
+                continuation.resume(throwing: error)
```
Comment (jbelkins) [inline Sources/ClientRuntime/Networking/Http/CRT/CRTClientEngine.swift:349]: good catch

File: Sources/ClientRuntime/Networking/Http/HttpClientConfiguration.swift [hunk 2]
```diff
@@ -9,11 +9,18 @@ import struct Foundation.TimeInterval
...
     ///
     /// If none is provided, the client will use default values based on the platform.
     public var connectTimeout: TimeInterval?
 
+    /// The timeout for socket, in seconds.
+    /// Sets maximum time to wait between two data packets.
+    /// Used to close stale connections that have no activity.
+    ///
+    /// If no value is provided, the defaut client won't have a socket timeout.
+    public var socketTimeout: TimeInterval?
```
Comment (jbelkins) [inline Sources/ClientRuntime/Networking/Http/HttpClientConfiguration.swift:22]: URLSession does not have separate settings for timeout on initial connection, and for idle time between data transmissions on an existing connection.

https://developer.apple.com/documentation/foundation/urlsessionconfiguration/1408259-timeoutintervalforrequest

seems to be the setting for both and the two cannot be set individually.
Our doc comments should make clear which settings apply to which HTTP client and which are ignored.

File: Sources/ClientRuntime/Networking/Http/URLSession/URLSessionConfiguration+HTTPClientConfiguration.swift [hunk 3]
```diff
@@ -13,8 +13,8 @@ extension URLSessionConfiguration {
 
     public static func from(httpClientConfiguration: HttpClientConfiguration) -> URLSessionConfiguration {
         let config = URLSessionConfiguration.default
-        if let connectTimeout = httpClientConfiguration.connectTimeout {
-            config.timeoutIntervalForRequest = connectTimeout
+        if let socketTimeout = httpClientConfiguration.socketTimeout {
+            config.timeoutIntervalForRequest = socketTimeout
```
Comment (jbelkins) [inline Sources/ClientRuntime/Networking/Http/URLSession/URLSessionConfiguration+HTTPClientConfiguration.swift:17]: See my note above.  How we transfer the HTTP config fields to specific clients should be made clear in the doc comments.

**Review by jbelkins — CHANGES_REQUESTED** [review]
Technically this looks good.  But we need to make sure through doc comments on the related config properties that customers understand what fields have what effect on the two supported HTTP clients.

File: Sources/ClientRuntime/Networking/Http/HttpClientConfiguration.swift (hunk 2 above)
Comment (sichanyoo) [reply Sources/ClientRuntime/Networking/Http/HttpClientConfiguration.swift:22]: Found a separate setting for initial connection timeout in URLRequest: https://developer.apple.com/documentation/foundation/urlrequest/2011509-timeoutinterval/

connectTimeout config on HttpClientConiguration is now set for URLRequest:timeoutInterval.

File: Sources/ClientRuntime/Networking/Http/URLSession/URLSessionConfiguration+HTTPClientConfiguration.swift (hunk 3 above)
Comment (sichanyoo) [reply Sources/ClientRuntime/Networking/Http/URLSession/URLSessionConfiguration+HTTPClientConfiguration.swift:17]: Both config fields are now used for same purposes in both clients, so I think existing comments do their job without exposing arbitrary details.

**Review by jbelkins — APPROVED** [review]

**Pull request merged**